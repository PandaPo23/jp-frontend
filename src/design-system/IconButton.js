import isNil from 'lodash/isNil';
import React from 'react';

import Flex from './Flex';
import Icon from './Icon';
import targetedProps from '../utils/targetedProps';
import Text from './Text';

const vert = (lp) => lp === 'top' || lp === 'bottom';
const before = (lp) => lp === 'left' || lp === 'top';

const IconButton = React.forwardRef(
  (
    {
      name,
      size = 24,
      bg = 'transparent',
      color = 'misc.muted',
      disabled = false,
      alignItems = 'center',
      border = 'none',
      label,
      children,
      labelPosition = 'right',
      labelMargin = 1,
      hoverable = true,
      justifyContent,
      labelUppercase = true,
      labelBold = true,
      as = 'button',
      ...props
    },
    ref
  ) => {
    const iconProps = targetedProps(props, 'icon');
    const labelProps = targetedProps(props, 'label');
    const justIcon = isNil(label) && !children;
    return (
      <Flex
        disabled={disabled}
        hoverable={hoverable}
        flexDirection={justIcon || vert(labelPosition) ? 'column' : 'row'}
        flexWrap="nowrap"
        nobr
        justifyContent={justifyContent || (justIcon && 'center')}
        alignItems={alignItems}
        border={border}
        color={color}
        bg={bg}
        as={as}
        ref={ref}
        {...props}
      >
        <Icon
          name={name}
          size={size}
          order={before(labelPosition) ? 2 : 1}
          {...iconProps}
        />
        {(children || !isNil(label)) && (
          <Text
            bg="transparent"
            disabled={disabled}
            order={before(labelPosition) ? 1 : 2}
            ml={labelPosition === 'right' && labelMargin}
            mr={labelPosition === 'left' && labelMargin}
            mt={labelPosition === 'bottom' && labelMargin}
            mb={labelPosition === 'top' && labelMargin}
            truncate
            nobr
            uppercase={labelUppercase}
            bold={labelBold}
            fontSize={3}
            {...labelProps}
          >
            {children || label}
          </Text>
        )}
      </Flex>
    );
  }
);

IconButton.displayName = 'IconButton';

export default IconButton;
