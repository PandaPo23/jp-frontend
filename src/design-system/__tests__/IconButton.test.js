import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import IconButton from '../IconButton';

describe('IconButton component', () => {
  it('should render as IconButton', () => {
    const wrapper = mount(<IconButton />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render IconButton with passed props', () => {
    const props = {
      name: 'menu',
    };
    expect(
      mount(<IconButton {...props} />)
        .find(IconButton)
        .prop('name')
    ).toBe('menu');
  });
  it('simulates click events', () => {
    const onClick = jest.fn();

    const button = mount(<IconButton onClick={onClick} />);
    expect(onClick).not.toBeCalled();
    button
      .find(IconButton)
      .first()
      .simulate('click');
    expect(onClick).toBeCalled();
  });
  // it('should render disabled IconButton', () => {
  //   const onClick = jest.fn();
  //   const button = mount(<IconButton disabled onClick={onClick} />);
  //   expect(onClick).not.toBeCalled();
  //   button
  //     .find(IconButton)
  //     .first()
  //     .simulate('click');
  //   expect(onClick).not.toBeCalled();
  // });
});
