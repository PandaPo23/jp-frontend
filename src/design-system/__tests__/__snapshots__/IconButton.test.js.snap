// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton component should render as IconButton 1`] = `
.emotion-4 {
  color: misc.muted;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: none;
  cursor: pointer;
}

.emotion-0 {
  min-width: 24px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 24px;
}

<IconButton>
  <Flex
    alignItems="center"
    as="button"
    bg="transparent"
    border="none"
    color="misc.muted"
    disabled={false}
    display="flex"
    flexDirection="column"
    flexWrap="nowrap"
    hoverable={true}
    justifyContent="center"
    nobr={true}
  >
    <button
      className="emotion-4"
      disabled={false}
    >
      <Icon
        name="check"
        order={1}
        size={24}
      >
        <Base
          className="emotion-0"
          name="check"
          order={1}
          size={24}
        >
          <Flex
            centered={true}
            className="emotion-0"
            display="flex"
            order={1}
          >
            <div
              className="emotion-1"
            >
              <svg
                className="emotion-0"
                name="check"
                order={1}
                size={24}
              />
            </div>
          </Flex>
        </Base>
      </Icon>
    </button>
  </Flex>
</IconButton>
`;
